test/api/model/model_test.dart

```diff
@@ -0,0 +1,50 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/api/model/model.dart';
+
+void main() {
+  group('user', () {
+    final Map<String, dynamic> baseJson = Map.unmodifiable({
+      'user_id': 123,
+      'delivery_email': 'name@example.com',
+      'email': 'name@example.com',
+      'full_name': 'A User',
+      'date_joined': '2023-04-28',
+      'is_active': true,
+      'is_owner': false,
+      'is_admin': false,
+      'is_guest': false,
+      'is_billing_admin': false,
+      'is_bot': false,
+      'role': 400,
+      'timezone': 'UTC',
+      'avatar_version': 0,
+      'profile_data': <String, dynamic>{},
+    });
+    User mkUser (Map<String, dynamic> specialJson) {
+      final json = Map.of(baseJson);
+      json.addAll(specialJson);
+      return User.fromJson(json);
+    }
+
+    test('delivery_email', () {
+      check(mkUser({'delivery_email': 'name@email.com'}).deliveryEmailStaleDoNotUse)
+        .equals('name@email.com');
+    });
+
+    test('profile_data', () {
+      check(mkUser({'profile_data': <String, dynamic>{}}).profileData).isNull();
+      check(mkUser({'profile_data': null}).profileData).isNull();
+      check(mkUser({'profile_data': {'1': {'value': 'foo'}}}).profileData)
+        .isNotNull().entries.single
+          ..has((s) => s.key, 'key').equals(1)
+          ..has((s) => s.value, 'value').isA<ProfileFieldUserData>();
+    });
+
+    test('is_system_bot', () {
+      assert(baseJson['is_system_bot'] == null); // could refactor
```

Comment:
> or as an explanation of why the assert is needed ("we could refactor how `baseJson` and `mkUser` work, and accidentally defeat this test")?

Yeah, that's why I included the assert. But I didn't state that reason explicitly (oops); my comment "could refactor" was like a TODO to fix an awkwardness with the test setup: the test wants `is_system_bot` to be absent, but it has no way to ensure that, using the available interface (`mkUser`'s `specialJson` arg). I think ideally it would have a way to do that. But I didn't say `TODO` because it won't really call out for a refactor until we change `baseJson`/`mkUser` in a way that trips up the `assert`, if we ever do.

> As a bonus, that makes this test no longer care about the internals of how `mkUser` does its job, in the form of the relationship of the latter to `baseJson`.

Sure, that sounds good. The awkwardness I mentioned is still there, but at least the `check` gives the same guarantee as the `assert` did: it makes sure `mkUser` won't include `isSystemBot` unless asked by the caller. (I guess it now also checks that `User.isSystemBot` can be null, which is redundant with the static types.)